exercises/circular-buffer/example.h

```diff
@@ -0,0 +1,54 @@
+#if !defined(CIRCULAR_BUFFER_H_)
+#define CIRCULAR_BUFFER_H_
+
+#include <vector>
+#include <stdexcept>
+
+namespace circular_buffer {
+
+template <typename ValueType>
+class circular_buffer {
+public:
+    circular_buffer(std::size_t capacity) 
+        : buffer_(capacity), head_(0), tail_(0) {}
+    ValueType read();
+    void write(ValueType item);
+    void overwrite(ValueType item);
+    void clear() { head_ = 0; tail_ = 0; }
+
+private:
+    std::vector<ValueType> buffer_;
+    std::size_t head_;
+    std::size_t tail_;
+
+    void insert(ValueType item);
+    bool is_empty() const { return head_ == tail_; }
+    bool is_full() const { return head_ == (tail_ + 1) % (buffer_.size() + 1); }
```

Comment:
There's a bug here, though it's not caught in the test suite. This test should pass: 

``` cpp
TEST_CASE("custom_test") 
{
    circular_buffer::circular_buffer<value_type> buffer(1);

    REQUIRE_NOTHROW(buffer.write(1));
    buffer.overwrite(2);
    REQUIRE_THROWS_AS(buffer.write(2), std::domain_error);
}
```

Detecting if it's full is tricky just using the head and tail pointers. See https://embeddedartistry.com/blog/2017/4/6/circular-buffers-in-cc which has an explanation.

Reply:
@patricksjackson thank you for pointing me the obvious ;)
`tail_ - head_ >= buffer_.size()`will solve it.

I have a question. Potentially, students may implement `overwrite()` method through `write()` which throws an exception, but `overwrite()` mustn't follow this logic. Should I modify test cases to prevent such scenario?

Reply:
I don't think just that will solve it. The problem (or one problem) is that `overwrite()` increments `head_` by constraining the value to valid indexes, but `head_` and `tail_` are incremented in other places without doing that. It really is not obvious, it's somewhat tricky to get right.


On `overwrite()` exceptions:

Writing `overwrite()` in terms of `write()` isn't necessarily incorrect. If you wanted `overwrite()` to be `noexcept` you can still make that guarantee by correctly handling the exception case. And it's not clear to me that `overwrite()` should even be `noexcept`, since it should probably throw if the buffer is size 0.

So no, I don't think it's necessary (though maybe I'm misunderstanding?). I am curious how you actually would test that though, because I can't think of a way to do it.